File: AOC_2022/Day 12/12.py
from typing import List, Tuple

class Solution:
    def __init__(self):
        self.grid = []
        self.start = (0, 0)
        self.finish = (0, 0)
    
    def init(self, input: List[str]):
        self.grid = [list(row) for row in input]
        for i in range(len(self.grid)):
            for j in range(len(self.grid[i])):
                if self.grid[i][j] == 'S':
                    self.start = (i, j)
        #print(len(self.grid), len(self.grid[0]))
        # for row in self.grid:
        #     print(row)

    def within_boundary(self, n: int, m: int, p: Tuple[int, int]) -> bool:
        if (0 <= p[0] and p[0] < n) and (0 <= p[1] and p[1] < m):
            return True
        #print(p)
        return False
    
    def part1(self):
        q = [(self.start, 0)]
        self.grid[self.start[0]][self.start[1]] = 'a'
        dp = [[-1 for _ in range(len(self.grid[0]))] for _ in range(len(self.grid))]
        dp[self.start[0]][self.start[1]] = 0
        while q:
            curr, curr_dist = q.pop(0)
            if self.grid[curr[0]][curr[1]] == 'E':
                self.finish = curr
                print(f"Part1: {dp[curr[0]][curr[1]]}")
                return
            ch = chr(ord(self.grid[curr[0]][curr[1]]) + 1)
            if self.within_boundary(len(self.grid), len(self.grid[0]), (curr[0]+1, curr[1])):
                if self.grid[curr[0]+1][curr[1]] == 'E':
                    if self.grid[curr[0]][curr[1]] >= 'y' and dp[curr[0]+1][curr[1]] == -1:
                        q.append(((curr[0]+1, curr[1]), curr_dist+1))
                        dp[curr[0]+1][curr[1]] = curr_dist+1
                else:
                    if self.grid[curr[0]+1][curr[1]] <= ch and dp[curr[0]+1][curr[1]] == -1:
                        q.append(((curr[0]+1, curr[1]), curr_dist+1))
                        dp[curr[0]+1][curr[1]] = curr_dist+1
            if self.within_boundary(len(self.grid), len(self.grid[0]), (curr[0]-1, curr[1])):
                if self.grid[curr[0]-1][curr[1]] == 'E':
                    if self.grid[curr[0]][curr[1]] >= 'y' and dp[curr[0]-1][curr[1]] == -1:
                        q.append(((curr[0]-1, curr[1]), curr_dist+1))
                        dp[curr[0]-1][curr[1]] = curr_dist+1
                else:
                    if self.grid[curr[0]-1][curr[1]] <= ch and dp[curr[0]-1][curr[1]] == -1:
                        q.append(((curr[0]-1, curr[1]), curr_dist+1))
                        dp[curr[0]-1][curr[1]] = curr_dist+1
            if self.within_boundary(len(self.grid), len(self.grid[0]), (curr[0], curr[1]+1)):
                if self.grid[curr[0]][curr[1]+1] == 'E':
                    if self.grid[curr[0]][curr[1]] >= 'y' and dp[curr[0]][curr[1]+1] == -1:
                        q.append(((curr[0], curr[1]+1), curr_dist+1))
                        dp[curr[0]][curr[1]+1] = curr_dist+1
                else:
                    if self.grid[curr[0]][curr[1]+1] <= ch and dp[curr[0]][curr[1]+1] == -1:
                        q.append(((curr[0], curr[1]+1), curr_dist+1))
                        dp[curr[0]][curr[1]+1] = curr_dist+1
            if self.within_boundary(len(self.grid), len(self.grid[0]), (curr[0], curr[1]-1)):
                if self.grid[curr[0]][curr[1]-1] == 'E':
                    if self.grid[curr[0]][curr[1]] >= 'y' and dp[curr[0]][curr[1]-1] == -1:
                        q.append(((curr[0], curr[1]-1), curr_dist+1))
                        dp[curr[0]][curr[1]-1] = curr_dist+1
                else:
                    if self.grid[curr[0]][curr[1]-1] <= ch and dp[curr[0]][curr[1]-1] == -1:
                        q.append(((curr[0], curr[1]-1), curr_dist+1))
                        dp[curr[0]][curr[1]-1] = curr_dist+1
    
    def part2(self):
        q = [(self.finish, 0)]
        self.grid[self.finish[0]][self.finish[1]] = 'z'
        dp = [[-1 for _ in range(len(self.grid[0]))] for _ in range(len(self.grid))]
        dp[self.finish[0]][self.finish[1]] = 0
        while q:
            curr, curr_dist = q.pop(0)
            if self.grid[curr[0]][curr[1]] == 'a':
                print(f"Part2: {dp[curr[0]][curr[1]]}")
                return
            ch = chr(ord(self.grid[curr[0]][curr[1]]) - 1)
            if self.within_boundary(len(self.grid), len(self.grid[0]), (curr[0]+1, curr[1])):
                if self.grid[curr[0]+1][curr[1]] == 'a':
                    if self.grid[curr[0]][curr[1]] <= 'b' and dp[curr[0]+1][curr[1]] == -1:
                        q.append(((curr[0]+1, curr[1]), curr_dist+1))
                        dp[curr[0]+1][curr[1]] = curr_dist+1
                else:
                    if self.grid[curr[0]+1][curr[1]] >= ch and dp[curr[0]+1][curr[1]] == -1:
                        q.append(((curr[0]+1, curr[1]), curr_dist+1))
                        dp[curr[0]+1][curr[1]] = curr_dist+1
            if self.within_boundary(len(self.grid), len(self.grid[0]), (curr[0]-1, curr[1])):
                if self.grid[curr[0]-1][curr[1]] == 'a':
                    if self.grid[curr[0]][curr[1]] <= 'b' and dp[curr[0]-1][curr[1]] == -1:
                        q.append(((curr[0]-1, curr[1]), curr_dist+1))
                        dp[curr[0]-1][curr[1]] = curr_dist+1
                else:
                    if self.grid[curr[0]-1][curr[1]] >= ch and dp[curr[0]-1][curr[1]] == -1:
                        q.append(((curr[0]-1, curr[1]), curr_dist+1))
                        dp[curr[0]-1][curr[1]] = curr_dist+1
            if self.within_boundary(len(self.grid), len(self.grid[0]), (curr[0], curr[1]+1)):
                if self.grid[curr[0]][curr[1]+1] == 'a':
                    if self.grid[curr[0]][curr[1]] <= 'b' and dp[curr[0]][curr[1]+1] == -1:
                        q.append(((curr[0], curr[1]+1), curr_dist+1))
                        dp[curr[0]][curr[1]+1] = curr_dist+1
                else:
                    if self.grid[curr[0]][curr[1]+1] >= ch and dp[curr[0]][curr[1]+1] == -1:
                        q.append(((curr[0], curr[1]+1), curr_dist+1))
                        dp[curr[0]][curr[1]+1] = curr_dist+1        
            if self.within_boundary(len(self.grid), len(self.grid[0]), (curr[0], curr[1]-1)):
                if self.grid[curr[0]][curr[1]-1] == 'a':
                    if self.grid[curr[0]][curr[1]] <= 'b' and dp[curr[0]][curr[1]-1] == -1:
                        q.append(((curr[0], curr[1]-1), curr_dist+1))
                        dp[curr[0]][curr[1]-1] = curr_dist+1
                else:
                    if self.grid[curr[0]][curr[1]-1] >= ch and dp[curr[0]][curr[1]-1] == -1:
                        q.append(((curr[0], curr[1]-1), curr_dist+1))
                        dp[curr[0]][curr[1]-1] = curr_dist+1

def main():
    # with open("input.txt") as f:
    #     lines = f.readlines()
    file = open("input.txt", "r")
    lines = file.read().splitlines()
    s = Solution()
    s.init(lines)
    s.part1()
    s.part2()

if __name__ == "__main__":
    main()

File: AOC_2022/Day 12/12.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution
{
    public:
        vector<string> grid;
        pair<int,int> start,finish;
        void init(ifstream& input){
            string line;
            while(getline(input,line)){
                grid.push_back(line);
            }
            for(int i = 0; i < grid.size();i++){
                for(int j = 0; j < grid[i].size();j++){
                    if(grid[i][j] == 'S'){
                        start.first = i;
                        start.second = j;
                    }
                }
            }
        }
        bool withinBoundary(int n, int m, pair<int,int> p){
            if((p.first >= 0 && p.first < n) && (p.second >= 0 && p.second < m))
                return true;
            return false;
        }
        void part1(){
            queue<pair<int,int>> q;
            q.push(start);
            grid[start.first][start.second] = 'a';
            vector<vector<int>> dp(grid.size(),vector<int>(grid[0].size(),-1));
            dp[start.first][start.second] = 0;
            while(!q.empty()){
                pair<int,int> curr = q.front();
                q.pop();
                if(grid[curr.first][curr.second] == 'E'){
                    finish = make_pair(curr.first,curr.second);
                    cout<<"Part1: "<<dp[curr.first][curr.second]<<"\n";
                    return;
                }
                char ch = grid[curr.first][curr.second] + 1;
                if(withinBoundary(grid.size(),grid[0].size(),make_pair(curr.first+1,curr.second))){
                    if(grid[curr.first+1][curr.second] == 'E'){
                        if(grid[curr.first][curr.second] >= 'y' && dp[curr.first+1][curr.second] == -1){
                            q.push(make_pair(curr.first+1,curr.second));
                            dp[curr.first+1][curr.second] = dp[curr.first][curr.second]+1;
                        }
                    }
                    else{
                        if(grid[curr.first+1][curr.second] <= ch && dp[curr.first+1][curr.second] == -1){
                            q.push(make_pair(curr.first+1,curr.second));
                            dp[curr.first+1][curr.second] = dp[curr.first][curr.second]+1;
                        }
                    }
                }      
                if(withinBoundary(grid.size(),grid[0].size(),make_pair(curr.first-1,curr.second))){
                    if(grid[curr.first-1][curr.second] == 'E'){
                        if(grid[curr.first][curr.second] >= 'y' && dp[curr.first-1][curr.second] == -1){
                            q.push(make_pair(curr.first-1,curr.second));
                            dp[curr.first-1][curr.second] = dp[curr.first][curr.second]+1;
                        }
                    }
                    else{
                        if(grid[curr.first-1][curr.second] <= ch && dp[curr.first-1][curr.second] == -1){
                            q.push(make_pair(curr.first-1,curr.second));
                            dp[curr.first-1][curr.second] = dp[curr.first][curr.second]+1;
                        }
                    }
                }
                if(withinBoundary(grid.size(),grid[0].size(),make_pair(curr.first,curr.second+1))){
                    if(grid[curr.first][curr.second+1] == 'E'){
                        if(grid[curr.first][curr.second] >= 'y' && dp[curr.first][curr.second+1] == -1){
                            q.push(make_pair(curr.first,curr.second+1));
                            dp[curr.first][curr.second+1] = dp[curr.first][curr.second]+1;
                        }
                    }
                    else{
                        if(grid[curr.first][curr.second+1] <= ch && dp[curr.first][curr.second+1] == -1){
                            q.push(make_pair(curr.first,curr.second+1));
                            dp[curr.first][curr.second+1] = dp[curr.first][curr.second]+1;
                        }
                    }
                }
                if(withinBoundary(grid.size(),grid[0].size(),make_pair(curr.first,curr.second-1))){
                    if(grid[curr.first][curr.second-1] == 'E'){
                        if(grid[curr.first][curr.second] >= 'y' && dp[curr.first][curr.second-1] == -1){
                            q.push(make_pair(curr.first,curr.second-1));
                            dp[curr.first][curr.second-1] = dp[curr.first][curr.second]+1;
                        }
                    }
                    else{
                        if(grid[curr.first][curr.second-1] <= ch && dp[curr.first][curr.second-1] == -1){
                            q.push(make_pair(curr.first,curr.second-1));
                            dp[curr.first][curr.second-1] = dp[curr.first][curr.second]+1;
                        }
                    }
                }  
            }
        }
        void part2(){
            queue<pair<int,int>> q;
            q.push(finish);
            grid[finish.first][finish.second] = 'z';
            vector<vector<int>> dp(grid.size(),vector<int>(grid[0].size(),-1));
            dp[finish.first][finish.second] = 0;
            while(!q.empty()){
                pair<int,int> curr = q.front();
                q.pop();
                if(grid[curr.first][curr.second] == 'a'){
                    cout<<"Part2: "<<dp[curr.first][curr.second]<<"\n";
                    return;
                }
                char ch = grid[curr.first][curr.second]-1;
                if(withinBoundary(grid.size(),grid[0].size(),make_pair(curr.first+1,curr.second))){
                    if(grid[curr.first+1][curr.second] == 'a'){
                        if(grid[curr.first][curr.second] <= 'b' && dp[curr.first+1][curr.second] == -1){
                            q.push(make_pair(curr.first+1,curr.second));
                            dp[curr.first+1][curr.second] = dp[curr.first][curr.second]+1;
                        }
                    }
                    else{
                        if((grid[curr.first+1][curr.second] >= ch) && dp[curr.first+1][curr.second] == -1){
                            q.push(make_pair(curr.first+1,curr.second));
                            dp[curr.first+1][curr.second] = dp[curr.first][curr.second]+1;
                        }
                    }
                    
                }      
                if(withinBoundary(grid.size(),grid[0].size(),make_pair(curr.first-1,curr.second))){
                    if(grid[curr.first-1][curr.second] == 'a'){
                        if(grid[curr.first][curr.second] <= 'b' && dp[curr.first-1][curr.second] == -1){
                            q.push(make_pair(curr.first-1,curr.second));
                            dp[curr.first-1][curr.second] = dp[curr.first][curr.second]+1;
                        }
                    }
                    else{
                        if(grid[curr.first-1][curr.second] >= ch && dp[curr.first-1][curr.second] == -1){
                            q.push(make_pair(curr.first-1,curr.second));
                            dp[curr.first-1][curr.second] = dp[curr.first][curr.second]+1;
                        }
                    }   
                }
                if(withinBoundary(grid.size(),grid[0].size(),make_pair(curr.first,curr.second+1))){
                    if(grid[curr.first][curr.second+1] == 'a'){
                        if(grid[curr.first][curr.second] <= 'b' && dp[curr.first][curr.second+1] == -1){
                            q.push(make_pair(curr.first,curr.second+1));
                            dp[curr.first][curr.second+1] = dp[curr.first][curr.second]+1;
                        }
                    }
                    else{
                        if(grid[curr.first][curr.second+1] >= ch && dp[curr.first][curr.second+1] == -1){
                            q.push(make_pair(curr.first,curr.second+1));
                            dp[curr.first][curr.second+1] = dp[curr.first][curr.second]+1;
                        }
                    }
                }
                if(withinBoundary(grid.size(),grid[0].size(),make_pair(curr.first,curr.second-1))){
                    if(grid[curr.first][curr.second-1] == 'a'){
                        if(grid[curr.first][curr.second] <= 'b' && dp[curr.first][curr.second-1] == -1){
                            q.push(make_pair(curr.first,curr.second-1));
                            dp[curr.first][curr.second-1] = dp[curr.first][curr.second]+1;
                        }
                    }
                    else{
                        if(grid[curr.first][curr.second-1] >= ch && dp[curr.first][curr.second-1] == -1){
                            q.push(make_pair(curr.first,curr.second-1));
                            dp[curr.first][curr.second-1] = dp[curr.first][curr.second]+1;
                        }
                    }
                }  
            }
        }

};


int main(){
    ifstream in("input.txt");
    if (!in.is_open())
    {
        cerr << "Could not open the file - '" << endl;
        return EXIT_FAILURE;
    }
    Solution s;
    s.init(in);
    s.part1();
    s.part2();
    return 0;
}
